File: Baekjoon/Class2/39.1874 스택 수열.java
import java.io.*;
import java.util.ArrayDeque;
import java.util.Deque;

class Main {
    public static void main(String[] args) throws IOException {
        int n;
        Deque<Integer> numDeque = new ArrayDeque<>();
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        n = Integer.parseInt(br.readLine());
        numDeque.push(0); // top 접근을 위한 더미 데이터
        int numToPush = 1;
        for(int i = 0; i < n; i++) { // n번 pop
            int numToPop = Integer.parseInt(br.readLine());
            while(numDeque.peek() < numToPop) { // pop할 숫자가 없으면
                numDeque.push(numToPush++); // push하기
                sb.append("+\n");
            }
            if(numDeque.peek() > numToPop) { // top이 pop할 숫자 초과(실패)
                sb.setLength(0);
                sb.append("NO");
                break;
            }
            numDeque.pop(); // pop할 숫자 찾음
            sb.append("-\n");
        }
        System.out.print(sb);
        br.close();
    }
}

File: Baekjoon/Class2/39.1874 스택 수열.py
from collections import deque
import sys
input = sys.stdin.readline

n = int(input())
num_deque = deque([0]) #top 접근을 위한 더미 데이터
out = []

num_to_push = 1
for _ in range(n): #n번 pop
    num_to_pop = int(input())
    while num_deque[-1] < num_to_pop: #pop할 숫자가 없으면
        num_deque.append(num_to_push) #push하기
        num_to_push += 1
        out.append("+")
    if num_deque[-1] > num_to_pop: #top이 pop할 숫자 초과(실패)
        out = ["NO"]
        break
    num_deque.pop() #pop할 숫자 찾음
    out.append("-")

print("\n".join(out))